### DISCUSSION TITLE
Debugging code on the hub

### DISCUSSION THREAD

**Discussion by jeagh** [pybricks/support#1028] [Q&A]
I've solved the problem that I had but it was a frustrating exercise because I was relying on print statements that wouldn't print.

The following is the pc to hub example and not the code I was working on. I've added a print statement so that I could see what was being received, but nothing was printed. What am I missing here?
```python
# NOTE: Run this program with the latest
# firmware provided via https://beta.pybricks.com/

from pybricks.pupdevices import Motor
from pybricks.parameters import Port
from pybricks.tools import wait

# Standard MicroPython modules
from usys import stdin, stdout
from uselect import poll

motor = Motor(Port.A)

# Optional: Register stdin for polling. This allows
# you to wait for incoming data without blocking.
keyboard = poll()
keyboard.register(stdin)

while True:

    # Optional: Check available input.
    while not keyboard.poll(0):
        # Optional: Do something here.
        wait(10)

    # Read three bytes.
    cmd = stdin.buffer.read(3)

    print(cmd) ###### This the added print statement and I also added a wait statement, just in case.

    # Decide what to do based on the command.
    if cmd == b"fwd":
        motor.dc(50)
    elif cmd == b"rev":
        motor.dc(-50)
    elif cmd == b"bye":
        break
    else:
        motor.stop()
    
    # Send a response.
    stdout.buffer.write(b"OK")

```

**Comment by dlech** [comment]
The `print()` function also uses `stdout` so it is going to your PC program the same as `stdout.buffer.write(b"OK")` does.

**Reply by jeagh** [reply]
Thank you. I've gone back to the project to check that I hadn't notice that the print statements were displayed on the pc, but they don't. I also tried stdout.write('test) and that's not displayed either. As I say, I've fixed the original problem but I'd still like to know why I cannot see the result of the print statements. I'll continue to research this.

It's possible that my changes to the example code have stopped the print statements and the stdout.write statements from being displayed.

**Reply by dlech** [reply]
It is up to the PC program to echo the received data or not.

I recently made a pull request to update this example code that might work more like you are expecting. You can find it [here for pc](https://github.com/pybricks/pybricks-projects/blob/150a5c1b625f1fdc34e975f6110f3d28c975f601/tutorials/wireless/hub-to-device/pc-communication/demo.py) and [here for hub](https://github.com/pybricks/pybricks-projects/blob/150a5c1b625f1fdc34e975f6110f3d28c975f601/tutorials/wireless/hub-to-device/pc-communication/main.py).

**Reply by jeagh** [reply]
Thank you again David,

My Internet connection has slowed to a crawl, however, I will try to connect to the links that you provided.

**Comment by jeagh** [comment]
The improved pc code does display the print statements from the hub, which is good.

My pc code was based on the revised code provided by Laurens which worked well. I'm now using the new revised hub and pc code but have run into a problem; the hub disconnects after receiving one byte from the pc. This is a snippet from the pc code where I'm trying to send bytes from a list instead of 'fwd' and 'rev'
```python
        # Send a few messages to the hub.

        line =  (0, 0), (3, 97), (53, 97), (53, 47), (3, 47), (3,  97), (0, 0)
        
        for i in range(len(line)):
            cmd = f'X{line[i][0]:03d} Y{line[i][1]:03d}'

           # ready_event.set()
            await send(cmd.encode())
            time.sleep(1)
            #ready_event.wait()

````
This is received from the hub; 
Received: bytearray(b"line =  b\'X000 Y000\'")
Received: bytearray(b'\r\n')
Received: bytearray(b'x =  0   y =  0\r\n')
Hub was disconnected.

What dumb beginner's error have I made?

**Reply by dlech** [reply]
It is difficult to guess what the problem could be without seeing both programs in full.

---

> ```python
> time.sleep(1)
> ```

This needs to be `await asyncio.sleep(1)`. Mixing blocking code and asyncio is a recipe for failure.

---

> ```python
>         for i in range(len(line)):
>             cmd = f'X{line[i][0]:03d} Y{line[i][1]:03d}'
> ```

More pythonic and easier to read would be:

```python
for x, y in line:
    cmd = f'X{x:03d} Y{y:03d}'
```

---

> ```python
>             #ready_event.wait()
> ```

Unless you have provided a different synchronization mechanism, commenting out the `read_event` code probably isn't a good idea.

**Reply by jeagh** [reply]
> This needs to be await asyncio.sleep(1). Mixing blocking code and asyncio is a recipe for failure

Thant was the problem, thank you again David.
